cadcdata/cadcdata/core.py

```diff
@@ -0,0 +1,418 @@
+#  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence                                                                                  
+#  General Public License for           Générale Publique GNU Affero
+#  more details.                        pour plus de détails.
+#                                       
+#  You should have received             Vous devriez avoir reçu une
+#  a copy of the GNU Affero             copie de la Licence Générale
+#  General Public License along         Publique GNU Affero avec
+#  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+#  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+#                                       <http://www.gnu.org/licenses/>.
+#
+#  $Revision: 4 $
+#
+# ***********************************************************************
+#
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import argparse
+import imp
+import logging
+import os
+import os.path
+import sys
+from StringIO import StringIO
+from datetime import datetime
+
+from cadcutils import net, util, exceptions
+from caom2.obs_reader_writer import ObservationReader, ObservationWriter
+from caom2.version import version as caom2_version
+from six.moves.urllib.parse import urlparse
+import zlib
+
+# from . import version as caom2repo_version
+from cadcdata import version
+
+__all__ = ['CadcDataClient']
+
+# TODO replace with SERVICE_URI when server supports it
+SERVICE_URL = 'www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/'
+# IVOA dateformat
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+DEFAULT_RESOURCE_ID = 'ivo://cadc.nrc.ca/data'
+ARCHIVE_STREAM_HTTP_HEADER = 'X-CADC-Stream'
+APP_NAME = 'cadc-data'
+
+READ_BLOCK_SIZE=8*1024
+
+logger = logging.getLogger(APP_NAME)
+
+class CadcDataClient:
+
+    """Class to access CADC archival data."""
+
+    logger = logging.getLogger('CadcDataClient')
+
+    def __init__(self, subject, resource_id=DEFAULT_RESOURCE_ID, host=None):
+        """
+        Instance of a CAOM2RepoClient
+        :param subject: the subject(user) performing the action
+        :type subject: Subject
+        :param resource_id: The identifier of the service resource (e.g 'ivo://cadc.nrc.ca/data')
+        :param host: Host server for the caom2repo service
+        """
+
+        self.resource_id = resource_id
+
+        # TODO This is just a temporary hack to be replaced with proper registry lookup functionaliy
+        resource_url = urlparse(resource_id)
+        self.host = host
+
+        agent = "cadc-data-client/{}".format(version.version)
+
+        self._repo_client = net.BaseWsClient(resource_id, subject,
+                                             agent, retry=True, host=self.host)
+        self.logger.info('Service URL: {}'.format(self._repo_client.base_url))
+
+
+    def get_file(self, archive, file_id, file=None, decompress=False,
+                 cutout=None, fhead=False, wcs=False, process_bytes=None):
```

Comment:
Oh yeah, that's even better. But in that case I will rename the argument destination, where destination can be a file name, file handle, buffer or anything that support open/close + write.

In spite of this,  the process_bytes argument is still useful, isn't it? I'm thinking of cases in which the file is too big to load into memory but you want to peek at the bytes as they flow in to extract info or process them on the fly.